File: BenKim/2022-09-15-product-of-array-except-self.py
# 🥲 time out solution 
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        result_dict = [0] * len(nums)
        for i, num in enumerate(nums):
            # 현재요소를 제외한 배열
            except_arr = nums[:i] + nums[i+1:]
            
            # 현재요소 제외한 모든값을 곱해서 결과배열에 저장
            result_dict[i] = reduce(lambda x, y: x * y, except_arr)
        return result_dict

File: BrianLee/Week1/2022-09-15-Product-of-Array-Except-Self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int allProduct = 1;
        int zeroCount = 0;
        for(int num : nums) {
            if(num == 0) zeroCount++;
            else allProduct *= num;

            if(zeroCount > 1) {
                allProduct = 0;
                break;
            }
        }

        int[] result = new int[nums.length];
        if(allProduct != 0) {
            for(int i = 0; i < nums.length; i++) {
                if(nums[i] == 0) result[i] = allProduct;
                else if(zeroCount == 1) result[i] = 0;
                else result[i] = allProduct / nums[i];
            }
        }
        return result;
    }
}